File: 4_sum.cpp
// Question Link - https://www.interviewbit.com/problems/4-sum/
// solution 

vector<vector<int> > Solution::fourSum(vector<int> &A, int B) {
    unordered_map<int,vector<pair<int,int>>> mp;
    for(int i=0; i<A.size(); i++)
    {
        for(int j=i+1; j<A.size(); j++)
        {
            mp[A[i]+A[j]].push_back(make_pair(i,j));
        }
    }
    set<vector<int>> sv;
    for(int i=0; i<A.size(); i++)
    {
        for(int j=i+1; j<A.size(); j++)
        {
            int summ = B-A[i]-A[j];
            if(mp.find(summ) != mp.end())
            {
                for(int k=0; k<mp[summ].size(); k++)
                {
                    if(i != mp[summ][k].first && i != mp[summ][k].second && j != mp[summ][k].first && j != mp[summ][k].second)
                    {
                        vector<int>v;
                        v.push_back(A[i]);v.push_back(A[j]);v.push_back(A[mp[summ][k].first]);v.push_back(A[mp[summ][k].second]);
                        sort(v.begin(),v.end());
                        sv.insert(v);
                    }
                }
            }
        }
    }
    vector<vector<int>> v;
    for(auto itr=sv.begin(); itr!=sv.end(); itr++)
    {
        v.push_back(*itr);
    }
    return v;
}

File: striver_sheet/arrays_part_IV/4_sum.py
# Leetcode solution
# Link : https://leetcode.com/problems/4sum/

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        x = []
        previ = 0
        i=0
        nums.sort()
        while i<len(nums):
            if previ != i and nums[previ] == nums[i]:
                i+=1
                continue
            elif previ != i and nums[previ] != nums[i]:
                previ = i
            j = i+1
            prevj = j
            while j < len(nums):
                if prevj != j and nums[prevj] == nums[j]:
                    j+=1
                    continue
                elif prevj != j and nums[prevj] != nums[j]:
                    prevj = j

                k = j+1
                prevk = k
                l = len(nums)-1
                prevl = l

                # print("I am coming here ??")

                while k<l:
                    if prevk != k and nums[prevk] == nums[k]:
                        k+=1
                        continue
                    elif prevk != k and nums[prevk] != nums[k]:
                        prevk = k

                    if prevl != l and nums[prevl] == nums[l]:
                        l-=1
                        continue
                    elif prevl != l and nums[prevl] != nums[l]:
                        prevl = l

                    summ = nums[i] + nums[j] + nums[k] + nums[l]
                    if  summ == target:
                        temp = [nums[i],nums[j],nums[k],nums[l]]
                        x.append(temp)
                        k+=1
                        l-=1
                        continue
                    elif summ < target:
                        k+=1
                        continue
                    elif summ > target:
                        l-=1
                        continue

                j+=1
            i+=1
        return x
